my_pyscf/mcpdft/laspdft.py

```diff
@@ -32,30 +33,53 @@ def get_h2eff(self, mo_coeff=None):
             eri = ao2mo.full(self.mol, mo_coeff, verbose=self.verbose,
                                 max_memory=self.max_memory)
         return eri
-
-def get_mcpdft_child_class(mc, ot, **kwargs):
+        
+def get_mcpdft_child_class(mc, ot, DoLASSI=False,  **kwargs):
     mc_doc = (mc.__class__.__doc__ or 'No docstring for MC-SCF parent method')
    
     class PDFT(_LASPDFT, mc.__class__):
         __doc__= mc_doc + '\n\n' + _LASPDFT.__doc__
         _mc_class = mc.__class__
-
+       
         def get_h2eff(self, mo_coeff=None):
             if self._in_mcscf_env: return mc.__class__.get_h2eff(self, mo_coeff=mo_coeff)
             else: return _LASPDFT.get_h2eff(self, mo_coeff=mo_coeff)
 
-        make_one_casdm1s=mc.__class__.make_casdm1s
-        make_one_casdm2=mc.__class__.make_casdm2
-        
+        # This code will change once state specific RDMs are available
+        def make_one_casdm1s(self, ci=None, state=0, **kwargs):
+            if self._in_mcscf_env: return mc.__class__.make_casdm1s
+            else: 
+                if self.roots_casdm1s is None: self.roots_casdm1s=self.states_make_casdm1s(ci=ci, **kwargs)
+                return self.roots_casdm1s[state, :, :, :]
+
+        def make_one_casdm2(self, ci=None, state=0, **kwargs):
+            if self._in_mcscf_env: return mc.__class__.make_casdm2
+            else:
+                if self.roots_casdm2 is None: self.roots_casdm2=self.states_make_casdm2(ci=ci, **kwargs)
+                return  self.roots_casdm2[state, :, :, :, :]
+
+
         # TODO: in pyscf-forge/pyscf/mcpdft/mcpdft.py::optimize_mcscf_, generalize the number
         # of return arguments. Then the redefinition below will be unnecessary
+        # TODO: Have to include the LASCI along with LASSCF
         def optimize_mcscf_(self, mo_coeff=None, ci0=None, **kwargs):
             '''Optimize the MC-SCF wave function underlying an MC-PDFT calculation.
             Has the same calling signature as the parent kernel method. '''
             with _mcscf_env(self):
                 self.e_mcscf, self.e_cas, self.ci, self.mo_coeff, self.mo_energy = \
-                    self._mc_class.kernel(self, mo_coeff, ci0=ci0, **kwargs)[:-2]
+                    self._mc_class.kernel(self, mo_coeff, ci0=ci0, **kwargs)[:-2] 
+                
+                if DoLASSI:
+                    e_roots, si = lassi.lassi(self)
```

Comment:
Can you cache `si` rather than `roots_casdm*`? It's more in the spirit of PySCF to only cache final results, not intermediate quantities like density matrices. It also is relevant to the issues of lines 52-53 and 58-59 above

Reply:
Yessssss!